### PR TITLE
[WIP] Nodemailer integration

### PR DESCRIPTION
**Pull request by rossoskull**
This is a PR for the issue #6 
Tasks completed : 
- Created a route which takes the username and the email ID as the parameters, encrypts them and sends an activation link to the user's email ID.
- Created a route that handles the activation link, and verifies if the link is correct or not.
- Make the form work with this route.
- After verification, adds the member to the respective link

### CODE REVIEW

File: createmail.js [hunk 1]
```diff
@@ -0,0 +1,5 @@
...
+};
+
+module.exports.createMail = createMail;
```
Comment (aashutoshrathi) [inline createmail.js]: Add new line and indent it under **\`**

File: server.js [hunk 2]
```diff
@@ -37,6 +41,64 @@ dict['2017'] = b17;
...
+    service: 'gmail',
+    auth: {
+      user: mail,
+      pass: password,
+    }
+  });
+
+  var mailOptions = {
+    from: '"IIITV Coding Club" <codingclub@iiitv.ac.in>',
+    to: id,
```
Comment (aashutoshrathi) [inline server.js]: Send email to self too.

File: package.json [hunk 3]
```diff
@@ -9,11 +9,12 @@
...
   "dependencies": {
+    "axios": "^0.18.0",
     "express": "^4.16.3",
-    "nodemailer": "^4.6.8",
+    "gh-account-exists": "^0.1.0",
     "jquery": "^3.3.1",
-    "axios": "^0.18.0",
-    "gh-account-exists": "^0.1.0"
+    "nodemailer": "^4.6.8",
+    "url-crypt": "^1.2.1"
```
Comment (aashutoshrathi) [inline package.json]: Also, add `node-rsa`

File: server.js [hunk 4]
```diff
@@ -17,6 +20,7 @@ const b17 = process.env.B17
 const b18 = process.env.B18
 const outs = process.env.OUTS
 
+const key = new rsa({b: 512});
```
Comment (singhpratyush) [inline server.js]:
```js
{ b: 512 }
```

Please add some code linter to the project.

File: server.js [hunk 5]
```diff
@@ -37,6 +41,64 @@ dict['2017'] = b17;
...
+// Testing 
+
+app.get('/verify/:base64', (req, res, next) => {
+  const encryptedData = req.params.base64;
+  var data;
+  const pass = true;
+  try {
+    data = urlcrypt.decryptObj(encryptedData);
+  } catch (e) {
+    res.send("Invalid Link.");
```
Comment (singhpratyush) [inline server.js]: Would this send a success response here? We should send some error response maybe (like bad request).

File: public/client.js [hunk 6]
```diff
@@ -63,11 +63,25 @@ $(function() {
     var username = $('#username').val()  
     var email = $('#email').val()
     console.log(username);
-    $.get('/add?' + $.param({username: username, email: email}), function() {
-      $('input').val('')
-      $('input').focus()
-      window.location.reload();
+    fetch('/sendmail/' + username + '/' + email)
```
Comment (singhpratyush) [inline public/client.js]: Same here, better use `.

File: server.js [hunk 7]
```diff
@@ -37,6 +41,64 @@ dict['2017'] = b17;
...
+
+app.get("/sendmail/:username/:id", (request, response, next) => {
+  const username = request.params.username;
+  const id = request.params.id;
+  const base64 = urlcrypt.cryptObj({
+    id: id,
+    username: username
+  });
+  
+  const verificationurl = 'http://localhost:3000/verify/' + base64;
```
Comment (singhpratyush) [inline server.js]: Will we still be using localhost here?

File: server.js [hunk 8]
```diff
@@ -37,6 +41,64 @@ dict['2017'] = b17;
...
+
+  var mailOptions = {
+    from: '"IIITV Coding Club" <codingclub@iiitv.ac.in>',
+    to: id,
+    //cc: mail,
+    subject: 'Invitation to join IIITV Organization on GitHub',
+    html: createMail.createMail(username, verificationurl),
+  };
+
+  transporter.sendMail(mailOptions, function(error, info){
```
Comment (singhpratyush) [inline server.js]: Shouldn't we `Promisify` this and `await` here? Or move the `next()` call after `200` response code is returned?

File: server.js (hunk 7 above)
Comment (rossoskull) [reply server.js]: 😅

File: public/client.js (hunk 6 above)
Comment (rossoskull) [reply public/client.js]: Okay

File: server.js (hunk 5 above)
Comment (rossoskull) [reply server.js]: This route is for the link which the user will get in the E-mail, so I just displayed this text when the link was tempered. And when the link is correct, it redirects to the teams page of  IIITV org.

File: createmail.js [hunk 9]
```diff
@@ -0,0 +1,5 @@
...
...x 3px 20px 2px grey;cursor: pointer;background-color: rgb(46, 196, 0);}#verify:active {box-shadow: 3px 3px 20px 0px grey;cursor: pointer;background-color: rgb(18, 77, 0);}.social-link {text-decoration: none;font-size: 18px;color: rgb(38, 102, 141);}</style></head><body><div id='root'><div id='header'><div id='brand-logo-cont' ><img id='brand-logo' src='https://scontent.fbom3-2.fna.fbcdn.net/v/t1.0-9/21616257_1209956105817094_7242516145331503157_n.png?_nc_cat=104&oh=c9f5720edd43b2b55f50d0f7e463e69d&oe=5C15672F' alt='IIITV Organization Logo'>  </div><h2>Welcome to IIITV Open Source Organization</h2></div><hr class='divider' /><main><h3>We're glad to have you with us, "+ username +"!</h3><p>Click on the button below to verify you E-Mail ID,<br>and you are good to go!</p><a href='"+ email +"'id='verify'>Verify E-Mail</a><p>With regards,<br>IIITV Open Source Organization</p></main><hr class='divider' /><footer><p>E-Mail us at codingclub@iiitvadodara.ac.in</p><h4>Find us on</h4><a target='_blank' class='social-link' href='https://www.facebook.com/iiitvcc/'>Facebook</a> |<a target='_blank' class='social-link' href='https://twitter.com/iiitvcc'>Twitter</a></footer>   </div></body></html>";
```
[the remark this answers is not in the document]
Comment (rossoskull) [reply createmail.js]: Okay, will do it

File: server.js (hunk 2 above)
Comment (rossoskull) [reply server.js]: Okay, will do it.

File: server.js (hunk 5 above)
Comment (singhpratyush) [reply server.js]: I was talking about this - 
```js
res.status(400).send('Invalid Link.');
```

File: package.json (hunk 3 above)
Comment (aashutoshrathi) [reply package.json]: Remove it.

**Comment by aashutoshrathi** [comment]
It still throws
```
Missing credentials for "PLAIN"
```

Refer to [this](https://github.com/nodemailer/nodemailer/issues/847)

**Comment by rossoskull** [comment]
@aashutoshrathi I have made some changes, can you try running it again?

File: createmail.js [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+                        <a target='_blank' class='social-link' href='https://www.facebook.com/iiitvcc/'>Facebook</a> |
+                        <a target='_blank' class='social-link' href='https://twitter.com/iiitvcc'>Twitter</a>
+                    </footer>
+                </div>
+            </body>
+        </html>    
+    `;
+};
+
+module.exports.createMail = createMail;
```
Comment (aashutoshrathi) [inline createmail.js]: Add newline

File: server.js [hunk 11]
```diff
@@ -61,30 +58,80 @@ app.get("/add", (request, response) => {
...
+    let url = "https://api.github.com/teams/" + dict[pref] + "/memberships/" + data.username + "?access_token=" + token;
+    console.log(url);
+    
+    axios.put(url)
+    .then(response => {
+      console.log(response.data.url);
+      resolve(200);
+    })
+    .catch(error => {
+      console.log("a\na\na\na\na\na\na\na\na\na\na\na\na\na\na\na\na\na\na\na\na\na\na\na\na\na\na\na\na\na\na\na\na\na\na\na\na\na\na\na\na\na\na\na\na\na\na\na\na\na\na\na\na\na\na\na\na\na\na\na\na\na\na\na\na\na\n");
```
Comment (aashutoshrathi) [inline server.js]: Why to print this much?

**Comment by aashutoshrathi** [comment]
Squash commits for once.

**Pull request merged**